File: src/planner.py
import numpy as np
from typing import List, Dict, Tuple


def solve_quintic(p0: float, v0: float, a0: float,
                  pT: float, vT: float, aT: float,
                  T: float) -> np.ndarray:
    """
    Solve a quintic polynomial coefficients for boundary conditions at t=0 and t=T.
    p(t) = a0 + a1 t + a2 t^2 + a3 t^3 + a4 t^4 + a5 t^5
    """
    T = float(T)
    M = np.array([
        [1, 0, 0, 0, 0, 0],
        [0, 1, 0, 0, 0, 0],
        [0, 0, 2, 0, 0, 0],
        [1, T, T**2, T**3, T**4, T**5],
        [0, 1, 2*T, 3*T**2, 4*T**3, 5*T**4],
        [0, 0, 2, 6*T, 12*T**2, 20*T**3],
    ], dtype=float)
    y = np.array([p0, v0, a0, pT, vT, aT], dtype=float)
    coeffs = np.linalg.solve(M, y)
    return coeffs


def sample_poly(coeffs: np.ndarray, t: np.ndarray) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
    """Sample polynomial, first and second derivatives at t points."""
    a0, a1, a2, a3, a4, a5 = coeffs
    p = a0 + a1*t + a2*t**2 + a3*t**3 + a4*t**4 + a5*t**5
    v = a1 + 2*a2*t + 3*a3*t**2 + 4*a4*t**3 + 5*a5*t**4
    a = 2*a2 + 6*a3*t + 12*a4*t**2 + 20*a5*t**3
    return p, v, a


def plan_quintic_xy(start: Dict[str, float], end: Dict[str, float],
                    T: float, N: int,
                    U_start: float) -> List[Dict[str, float]]:
    """
    Plan a 2D path using independent quintic polynomials in x and y.
    - start: {x, y, psi} psi in radians
    - end: {x, y, psi} psi in radians
    - U_start: initial speed magnitude (m/s)
    Returns list of {t, x, y, psi}
    """
    x0, y0, psi0 = float(start['x']), float(start['y']), float(start['psi'])
    xT, yT, psiT = float(end['x']), float(end['y']), float(end['psi'])

    # Initial/terminal velocities projected onto x/y axes
    vx0 = U_start * np.cos(psi0)
    vy0 = U_start * np.sin(psi0)
    # Terminal velocity magnitude uses same U_start for simplicity here
    vxT = U_start * np.cos(psiT)
    vyT = U_start * np.sin(psiT)

    ax0 = 0.0
    ay0 = 0.0
    axT = 0.0
    ayT = 0.0

    cx = solve_quintic(x0, vx0, ax0, xT, vxT, axT, T)
    cy = solve_quintic(y0, vy0, ay0, yT, vyT, ayT, T)

    t = np.linspace(0.0, T, int(N))
    x, vx, _ = sample_poly(cx, t)
    y, vy, _ = sample_poly(cy, t)
    psi = np.arctan2(vy, vx)

    plan = [
        {
            't': float(tt),
            'x': float(xx),
            'y': float(yy),
            'psi': float(ppsi),
        }
        for tt, xx, yy, ppsi in zip(t, x, y, psi)
    ]
    return plan


def _wrap_pi(a: float) -> float:
    return float((a + np.pi) % (2.0 * np.pi) - np.pi)


def _wrap_2pi(a: float) -> float:
    a = float(a % (2.0 * np.pi))
    return a


def plan_circle_arc(start: Dict[str, float], end: Dict[str, float],
                    T: float, N: int,
                    U_start: float) -> List[Dict[str, float]]:
    """
    以起点/终点位置与航向角生成单圆弧参考轨迹。
    - start, end: {x, y, psi}，psi 为弧度（tangent 方位）
    - 选择与起点航向相切的圆心方向（左/右），在两者中挑选使终点航向更贴近的方向。
    - 若几何退化（半径无效或近似直线），退化为直线插值（psi 为切线方向）。

    返回 {t, x, y, psi} 列表，t 依据传入 T 均匀分布。
    """
    x0, y0, psi0 = float(start['x']), float(start['y']), float(start['psi'])
    x1, y1, psi1 = float(end['x']), float(end['y']), float(end['psi'])

    p0 = np.array([x0, y0], dtype=float)
    p1 = np.array([x1, y1], dtype=float)
    d = p1 - p0
    dnorm = float(np.hypot(d[0], d[1]))
    if dnorm < 1e-6:
        # 起终点重合：输出一个常值点
        t = np.linspace(0.0, max(1e-3, float(T)), int(N))
        return [{
            't': float(tt), 'x': x0, 'y': y0, 'psi': float(psi0)
        } for tt in t]

    def try_dir(dir_s: int) -> Tuple[bool, Dict[str, float]]:
        # 圆心方向：起点航向的法线（左: +1，对应 CCW；右: -1，对应 CW）
        u0 = np.array([np.cos(psi0 + dir_s * np.pi / 2.0),
                       np.sin(psi0 + dir_s * np.pi / 2.0)], dtype=float)
        denom = float(np.dot(d, u0))
        if abs(denom) < 1e-8:
            return False, {}
        r = float((dnorm * dnorm) / (2.0 * denom))
        if not np.isfinite(r) or r <= 1e-6:
            return False, {}
        C = p0 + r * u0
        # 起终点的径向角
        theta0 = float(np.arctan2(y0 - C[1], x0 - C[0]))
        theta1 = float(np.arctan2(y1 - C[1], x1 - C[0]))
        if dir_s > 0:
            # CCW：角度递增
            dtheta = _wrap_2pi(theta1 - theta0)
        else:
            # CW：角度递减
            dtheta = -_wrap_2pi(theta0 - theta1)
        psi1_pred = float(theta1 + dir_s * np.pi / 2.0)
        err_end = abs(_wrap_pi(psi1_pred - psi1))
        return True, {
            'dir': dir_s,
            'C': C,
            'r': r,
            'theta0': theta0,
            'theta1': theta1,
            'dtheta': dtheta,
            'err_end': err_end,
        }

    okL, candL = try_dir(+1)
    okR, candR = try_dir(-1)
    cand = None
    if okL and okR:
        cand = candL if candL['err_end'] <= candR['err_end'] else candR
    elif okL:
        cand = candL
    elif okR:
        cand = candR
    else:
        # 几何退化：退化为直线插值
        t = np.linspace(0.0, max(1e-3, float(T)), int(N))
        xs = np.linspace(x0, x1, int(N))
        ys = np.linspace(y0, y1, int(N))
        # 直线切线方向
        psi_line = float(np.arctan2(y1 - y0, x1 - x0))
        return [{
            't': float(tt), 'x': float(xx), 'y': float(yy), 'psi': float(psi_line)
        } for tt, xx, yy in zip(t, xs, ys)]

    C = np.array(cand['C'], dtype=float)
    r = float(cand['r'])
    dir_s = int(cand['dir'])
    theta0 = float(cand['theta0'])
    dtheta = float(cand['dtheta'])

    # 采样角度与时间
    t = np.linspace(0.0, max(1e-3, float(T)), int(N))
    s = np.linspace(0.0, 1.0, int(N))
    thetas = theta0 + s * dtheta
    xs = C[0] + r * np.cos(thetas)
    ys = C[1] + r * np.sin(thetas)
    psis = thetas + dir_s * np.pi / 2.0

    return [{
        't': float(tt), 'x': float(xx), 'y': float(yy), 'psi': float(pp)
    } for tt, xx, yy, pp in zip(t, xs, ys, psis)]

File: src_cpp/src/planner.cpp
// 轨迹规划实现：五次多项式与单圆弧插值
#include "planner.hpp"
#include <cmath>
#include <vector>
#include <array>
#include <algorithm>

namespace fourws {

static inline double wrap_pi(double a) {
    const double two_pi = 2.0 * M_PI;
    double x = std::fmod(a + M_PI, two_pi);
    if (x < 0) x += two_pi;
    return x - M_PI;
}

static inline double wrap_2pi(double a) {
    const double two_pi = 2.0 * M_PI;
    double x = std::fmod(a, two_pi);
    if (x < 0) x += two_pi;
    return x;
}

std::vector<double> solve_quintic(double p0, double v0, double a0,
                                  double pT, double vT, double aT,
                                  double T) {
    T = (T <= 0.0) ? 1e-6 : T;
    // 6x6 线性方程求系数 a0..a5
    const double T2 = T*T;
    const double T3 = T2*T;
    const double T4 = T3*T;
    const double T5 = T4*T;

    // 矩阵行列式求解（显式求逆不必要，这里用 Cramer/直接构造解）
    // 使用标准推导：
    // a0=p0, a1=v0, a2=a0/2
    // a3=(20*(pT-p0) - (8*vT+12*v0)*T - (3*a0-aT)*T^2) / (2*T^3)
    // a4=(30*(p0-pT) + (14*vT+16*v0)*T + (3*a0-2*aT)*T^2) / (2*T^4)
    // a5=(12*(pT-p0) - (6*vT+6*v0)*T - (a0-aT)*T^2) / (2*T^5)
    std::vector<double> a(6);
    a[0] = p0;
    a[1] = v0;
    a[2] = a0 / 2.0;
    a[3] = (20.0*(pT - p0) - (8.0*vT + 12.0*v0)*T - (3.0*a0 - aT)*T2) / (2.0*T3);
    a[4] = (30.0*(p0 - pT) + (14.0*vT + 16.0*v0)*T + (3.0*a0 - 2.0*aT)*T2) / (2.0*T4);
    a[5] = (12.0*(pT - p0) - (6.0*vT + 6.0*v0)*T - (a0 - aT)*T2) / (2.0*T5);
    return a;
}

void sample_poly(const std::vector<double>& coeffs,
                 const std::vector<double>& t,
                 std::vector<double>& p,
                 std::vector<double>& v,
                 std::vector<double>& a) {
    const double a0 = coeffs[0];
    const double a1 = coeffs[1];
    const double a2 = coeffs[2];
    const double a3 = coeffs[3];
    const double a4 = coeffs[4];
    const double a5 = coeffs[5];
    const size_t N = t.size();
    p.resize(N); v.resize(N); a.resize(N);
    for (size_t i = 0; i < N; ++i) {
        const double tt = t[i];
        const double tt2 = tt*tt;
        const double tt3 = tt2*tt;
        const double tt4 = tt3*tt;
        const double tt5 = tt4*tt;
        p[i] = a0 + a1*tt + a2*tt2 + a3*tt3 + a4*tt4 + a5*tt5;
        v[i] = a1 + 2.0*a2*tt + 3.0*a3*tt2 + 4.0*a4*tt3 + 5.0*a5*tt4;
        a[i] = 2.0*a2 + 6.0*a3*tt + 12.0*a4*tt2 + 20.0*a5*tt3;
    }
}

std::vector<PlanPoint> plan_quintic_xy(const Pose2D& start,
                                       const Pose2D& end,
                                       double T,
                                       int N,
                                       double U_start) {
    const double x0 = start.x, y0 = start.y, psi0 = start.psi;
    const double xT = end.x,   yT = end.y,   psiT = end.psi;
    const double vx0 = U_start * std::cos(psi0);
    const double vy0 = U_start * std::sin(psi0);
    const double vxT = U_start * std::cos(psiT);
    const double vyT = U_start * std::sin(psiT);

    const auto cx = solve_quintic(x0, vx0, 0.0, xT, vxT, 0.0, T);
    const auto cy = solve_quintic(y0, vy0, 0.0, yT, vyT, 0.0, T);

    std::vector<double> t(N);
    for (int i = 0; i < N; ++i) t[i] = (N==1) ? 0.0 : (T * double(i) / double(N-1));
    std::vector<double> x, vx, ax;
    std::vector<double> y, vy, ay;
    sample_poly(cx, t, x, vx, ax);
    sample_poly(cy, t, y, vy, ay);

    std::vector<PlanPoint> plan; plan.reserve(N);
    for (int i = 0; i < N; ++i) {
        const double psi = std::atan2(vy[i], vx[i]);
        plan.push_back(PlanPoint{t[i], x[i], y[i], psi});
    }
    return plan;
}

std::vector<PlanPoint> plan_circle_arc(const Pose2D& start,
                                       const Pose2D& end,
                                       double T,
                                       int N,
                                       double /*U_start*/) {
    const double x0 = start.x, y0 = start.y, psi0 = start.psi;
    const double x1 = end.x,   y1 = end.y,   psi1 = end.psi;
    const double dx = x1 - x0, dy = y1 - y0;
    const double dnorm = std::hypot(dx, dy);
    if (dnorm < 1e-6) {
        std::vector<PlanPoint> out; out.reserve(N);
        const double Tuse = std::max(1e-3, T);
        for (int i = 0; i < N; ++i) {
            const double tt = (N==1) ? 0.0 : (Tuse * double(i) / double(N-1));
            out.push_back(PlanPoint{tt, x0, y0, psi0});
        }
        return out;
    }

    auto try_dir = [&](int dir_s, bool& ok, double& r, std::array<double,2>& C,
                       double& theta0, double& theta1, double& dtheta, double& err_end) {
        const double nx = std::cos(psi0 + dir_s * M_PI / 2.0);
        const double ny = std::sin(psi0 + dir_s * M_PI / 2.0);
        const double denom = dx*nx + dy*ny;
        if (std::abs(denom) < 1e-8) { ok = false; return; }
        r = (dnorm*dnorm) / (2.0 * denom);
        if (!std::isfinite(r) || r <= 1e-6) { ok = false; return; }
        C = {x0 + r*nx, y0 + r*ny};
        theta0 = std::atan2(y0 - C[1], x0 - C[0]);
        theta1 = std::atan2(y1 - C[1], x1 - C[0]);
        if (dir_s > 0) {
            dtheta = wrap_2pi(theta1 - theta0);
        } else {
            dtheta = -wrap_2pi(theta0 - theta1);
        }
        const double psi1_pred = theta1 + dir_s * M_PI / 2.0;
        err_end = std::abs(wrap_pi(psi1_pred - psi1));
        ok = true;
    };

    bool okL=false, okR=false; double rL=0, rR=0; std::array<double,2> CL{}, CR{};
    double th0L=0, th1L=0, dthL=0, errL=0;
    double th0R=0, th1R=0, dthR=0, errR=0;
    try_dir(+1, okL, rL, CL, th0L, th1L, dthL, errL);
    try_dir(-1, okR, rR, CR, th0R, th1R, dthR, errR);
    bool useL = false; bool useR = false;
    if (okL && okR) {
        useL = (errL <= errR); useR = !useL;
    } else if (okL) {
        useL = true;
    } else if (okR) {
        useR = true;
    } else {
        // 退化为直线插值
        std::vector<PlanPoint> out; out.reserve(N);
        const double psi_line = std::atan2(dy, dx);
        const double Tuse = std::max(1e-3, T);
        for (int i = 0; i < N; ++i) {
            const double s = (N==1) ? 0.0 : double(i) / double(N-1);
            const double tt = (N==1) ? 0.0 : (Tuse * s);
            const double xx = x0 + s*dx;
            const double yy = y0 + s*dy;
            out.push_back(PlanPoint{tt, xx, yy, psi_line});
        }
        return out;
    }

    const bool Lsel = useL;
    const double r = Lsel ? rL : rR;
    const auto& C = Lsel ? CL : CR;
    const double theta0 = Lsel ? th0L : th0R;
    const double dtheta = Lsel ? dthL : dthR;
    const int dir_s = Lsel ? +1 : -1;

    // 采样角度与时间
    std::vector<PlanPoint> out; out.reserve(N);
    const double Tuse = std::max(1e-3, T);
    for (int i = 0; i < N; ++i) {
        const double s = (N==1) ? 0.0 : double(i) / double(N-1);
        const double tt = (N==1) ? 0.0 : (Tuse * s);
        const double theta = theta0 + s * dtheta;
        const double xx = C[0] + r * std::cos(theta);
        const double yy = C[1] + r * std::sin(theta);
        const double psi = theta + dir_s * M_PI / 2.0;
        out.push_back(PlanPoint{tt, xx, yy, psi});
    }
    return out;
}

} // namespace fourws